cirq/_compat_test.py

```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
             ...
 
         # pylint: enable=unused-variable
+
+
+def _from_parent_import_deprecated():
+    from cirq.testing._compat_test_data import fake_a
+
+    assert fake_a
+
+
+def _import_deprecated_assert_sub():
+    import cirq.testing._compat_test_data.fake_a
+
+    assert cirq.testing._compat_test_data.fake_a.module_b
+
+
+def _from_deprecated_import_sub():
+    from cirq.testing._compat_test_data.fake_a import module_b
+
+    assert module_b
+
+
+def _import_deprecated_first_new_second():
+    """To ensure that module_a gets initialized only once."""
+    from cirq.testing._compat_test_data.fake_a import module_b
+
+    assert module_b
+    from cirq.testing._compat_test_data.module_a import module_b
+
+    assert module_b
```

Comment:
How does this enforce single-initialization?  Same for the following several tests.

Reply:
Oooops, I think what happened sometimes is that I pressed organize imports in PyCharm, that optimizes away unnecessary imports. I'll go back and fix all these. Thanks!

Reply:
Now that I'm looking at it - it does the job well, but probably could use an explanatory comment.

The way it works is through enforcing the count of info logs in `_test_deprecated_module_inner`: 

```
 with cirq.testing.assert_logs(
        "init:compat_test_data",
        "init:module_a",
        min_level=logging.INFO,
        max_level=logging.INFO,
        count=2,
    ):
```

It forces setting up the loader on the spec. Let's comment out this line in the `find_spec` method: 

```
spec.loader = DeprecatedModuleLoader(spec.loader, fullname, new_fullname)
```
In case of the following user code

```
from cirq.testing._compat_test_data.fake_a import module_b
from cirq.testing._compat_test_data.module_a import module_b
```

The importlib machinery first finds `fake_a` with our `DeprecatedModuleFinder` as `module_a`, it sets up a default Loader, which loads the module and places is the reference in `sys.modules['fake_a']`. 
Then, the second line executes again, finds `module_a` (which is NOT in the module cache at the moment), sets up a default Loader, which loads the module again, and places the reference in `sys.modules['module_a']`. 

The assertion error I get is 

```
 AssertionError: Expected 2 log message but got 3. Log messages: ['init:compat_test_data', 'init:module_a', 'init:module_a']
```

This is the double execution problem with these references. The `DeprecatedModuleLoader` does the "atomic" (not really as it's not prepared for multithreading...) update of the two references.